File: 528. Random Pick with Weight/solution.java
class Solution {
    private int[] sum;
    private Random random;
    private Map<Integer, Integer> map;
    public Solution(int[] w) {
        sum = new int[w.length];
        for (int i = 0; i < w.length; i++) {
            sum[i] = w[i] + (i == 0 ? 0 : sum[i - 1]);
        }

        // map = new HashMap<>();
        // int prev = 1;
        // for (int i = 0; i < sum.length; i++) {
        //     int val = sum[i];
        //     for (int j = prev; j <= val; j++) {
        //         // System.out.println("key: " + j + " value: " + i);
        //         map.put(j, i);
        //     }
        //     prev = val + 1;
        // }
        
        random = new Random();
    }
    
    public int pickIndex() {
        // [1, sum[sum.length - 1]]
        int index = random.nextInt(sum[sum.length - 1]) + 1;
        int b = 0; int e = sum.length - 1;
        while (b < e) {
            int mid = b + (e - b) / 2;
            if (sum[mid] == index) {
                return mid;
            } else if (sum[mid] < index) {
                b = mid + 1;
            } else {
                e = mid;
            }
        }
        return b;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

File: 528. Random Pick with Weight/solution.py
from bisect import bisect_left
import random
class Solution:

    def __init__(self, w: List[int]):
        self.prefix_sum = [0] * len(w)
        previous = 0
        for i in range(len(w)):
            self.prefix_sum[i] = w[i] + previous
            previous = self.prefix_sum[i]
        # [1, 3, 4,] -> [1, 4, 8]
        # generate a random number between 1 and prefix_sum[-1] and figure out the insert position, e.g 2
        # should fall in between of 1 and 3, thus, should return 3's index which is 2.

    def pickIndex(self) -> int:
        x = random.randint(1, self.prefix_sum[-1])
        # Find the leftmost value greater or equal to x
        i = bisect_left(self.prefix_sum, x)
        # i indicates the insertion position
        return i
        
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()
